speed-dreams/source-2.2.3/src/modules/userinterface/legacymenu/mainscreens/DatabaseConnectionCheck.cpp

```diff
@@ -0,0 +1,145 @@
+    strcpy_s(p_dbSettings.Username, SETTINGS_NAME_LENGTH, GfParmGetStr(p_param, PRM_USERNAME, GFMNU_ATTR_TEXT, nullptr));
+    strcpy_s(p_dbSettings.Password, SETTINGS_NAME_LENGTH, GfParmGetStr(p_param, PRM_PASSWORD, GFMNU_ATTR_TEXT, nullptr));
+    strcpy_s(p_dbSettings.Address, SETTINGS_NAME_LENGTH, GfParmGetStr(p_param, PRM_ADDRESS, GFMNU_ATTR_TEXT, nullptr));
+    strcpy_s(p_control.PortString, SETTINGS_NAME_LENGTH, GfParmGetStr(p_param, PRM_PORT, GFMNU_ATTR_TEXT, nullptr));
+    strcpy_s(p_dbSettings.Schema, SETTINGS_NAME_LENGTH, GfParmGetStr(p_param, PRM_SCHEMA, GFMNU_ATTR_TEXT, nullptr));
+    p_dbSettings.UseSSL = GfuiMenuControlGetBoolean(p_param, PRM_SSL, GFMNU_ATTR_CHECKED, false);
+    strcpy_s(p_dbSettings.CACertFilePath, SETTINGS_NAME_LENGTH, GfParmGetStr(p_param, PRM_CERT, GFMNU_ATTR_CA_CERT, nullptr));
+    strcpy_s(p_dbSettings.PublicCertFilePath, SETTINGS_NAME_LENGTH, GfParmGetStr(p_param, PRM_CERT, GFMNU_ATTR_PUBLIC_CERT, nullptr));
+    strcpy_s(p_dbSettings.PrivateCertFilePath, SETTINGS_NAME_LENGTH, GfParmGetStr(p_param, PRM_CERT, GFMNU_ATTR_PRIVATE_CERT, nullptr));
+
+    const char* filePath = GfParmGetStr(p_param, PRM_CERT, GFMNU_ATTR_CA_CERT, nullptr);
+    if (filePath)
+    {
+        strcpy_s(p_dbSettings.CACertFilePath, SETTINGS_NAME_LENGTH, filePath);
+    }
+    const char* filePath2 = GfParmGetStr(p_param, PRM_CERT, GFMNU_ATTR_PUBLIC_CERT, nullptr);
+    if (filePath2)
+    {
+        strcpy_s(p_dbSettings.PublicCertFilePath, SETTINGS_NAME_LENGTH, filePath2);
+    }
+    const char* filePath3 = GfParmGetStr(p_param, PRM_CERT, GFMNU_ATTR_PRIVATE_CERT, nullptr);
+    if (filePath3)
+    {
+        strcpy_s(p_dbSettings.PrivateCertFilePath, SETTINGS_NAME_LENGTH, filePath3);
+    }
+}
+
+/// @brief Loads the user menu settings from the local config file or the default values will be loaded
+void LoadDBSettings(void* p_scrHandle, DatabaseSettings& p_dbSettings, tDbControlSettings& p_control)
+{
+    // Retrieves the saved user xml file, if it doesn't exist the default values will be loaded
+    std::string strPath("config/DatabaseSettingsMenu.xml");
+    char buf[512];
+    sprintf(buf, "%s%s", GfLocalDir(), strPath.c_str());
+    if (GfFileExists(buf))
+    {
+        void* param = GfParmReadFile(buf, GFPARM_RMODE_STD);
+        // Initialize settings with the retrieved xml file
+        LoadConfigSettings(param, p_dbSettings, p_control);
+        return;
+    }
+    LoadDefaultSettings(p_scrHandle, p_dbSettings, p_control);
+}
+/// @brief                    The async function to check if a connection can be established between speed dreams and the database
+/// @param  p_scrHandle       The screen handle for writing on the screen
+/// @param  p_dbStatusControl The status control handle to write letters to the screen
+/// @param  p_dbSettings      The settings of the database
+void AsyncCheckConnection(void* p_scrHandle, int p_dbStatusControl, tDatabaseSettings p_dbSettings)
```

Comment:
Can you safely write async to the menu? What happens when you write when it's rendering?

Reply:
@KungFuDonkey I am not sure about this as Sietze implemented this hopefully you can tell more about this